internal/provider/resources/block.go

```diff
@@ -306,9 +308,121 @@ func (r *BlockResource) Read(ctx context.Context, req resource.ReadRequest, resp
 }
 
 // Update updates the resource and sets the updated Terraform state on success.
-//
-//nolint:revive // TODO: remove this comment when method is implemented
 func (r *BlockResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan BlockResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	blockTypeClient, err := r.client.BlockTypes(plan.AccountID.ValueUUID(), plan.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block Types", err))
+
+		return
+	}
+
+	blockSchemaClient, err := r.client.BlockSchemas(plan.AccountID.ValueUUID(), plan.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block Schema", err))
+
+		return
+	}
+
+	blockDocumentClient, err := r.client.BlockDocuments(plan.AccountID.ValueUUID(), plan.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block Document", err))
+
+		return
+	}
+
+	blockType, err := blockTypeClient.GetBySlug(ctx, plan.TypeSlug.ValueString())
```

Comment:
ok here's an interesting thing that came up - so the PATCH endpoint accepts a block_schema_id (meaning it's a modifiable attribute).  block_schemas are a "version" of a type

but what if the type_slug is modified in the TF resource?  then we'll get a failure like so - 

```
Terraform will perform the following actions:

  # prefect_block.lambda will be updated in-place
  ~ resource "prefect_block" "lambda" {
        id           = "adf48aa0-318d-40d1-aca0-c897f05cc22b"
        name         = "lambda"
      ~ type_slug    = "lambda-function" -> "secret"
      ~ updated      = "2024-06-06T23:40:59Z" -> (known after apply)
        # (3 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
prefect_block.lambda: Modifying... [id=adf48aa0-318d-40d1-aca0-c897f05cc22b]
╷
│ Error: Error during update Block Document
│
│   with prefect_block.lambda,
│   on main.tf line 19, in resource "prefect_block" "lambda":
│   19: resource "prefect_block" "lambda" {
│
│ Could not update Block Document, unexpected error: status code 400 Bad Request, error={"detail":"Must migrate block document to a block schema of the same
│ block type."}
```

we can find out, but im guessing the intent of the endpoint was to allow changing schema versions _**of the same block type**_, but we shouldnt' allow _**changing the type itself**_

so there's two things here:
1. should we prevent users from changing the `type_slug`? i think we should - but since the Schema object is shared, we may need to enforce this in this method (by comparing the `plan.TypeSlug == state.TypeSlug` => exit early if they do)
2. should we allow users modifying the schema of the same type? im not sure we need to (eg. the UI does not expose this).  also since we take the "[0]" schema from `blockSchemaClient.List`, we're effectively applying the latest schema on every update call

Reply:
> should we prevent users from changing the type_slug? i think we should - but since the Schema object is shared, we may need to enforce this in this method (by comparing the plan.TypeSlug == state.TypeSlug => exit early if they do)

@parkedwards good call-out. I started looking into this but haven't yet found a way to work with `state` and `plan` in the same place 🤔 Am I missing something?

I found a reference to [DiffSuppressFunc](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-behaviors#diffsuppressfunc), which isn't what we want, but does allow you to work with `old` and `new` values to compare them.

I'll keep looking but if you know of a way off the top of your head just lemme know 👍🏼

Reply:
Maybe [RequiresReplaceIf](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier#RequiresReplaceIf) could help here.

Reply:
take a peek at the `service_account.go` resource's Update() method - i had to do some similar kinds of `plan vs. state` checks.  the high level is that you can just invoke `req.Plan.Get()` or `req.State.Get()`, depending on what you want to draw from

https://github.com/PrefectHQ/terraform-provider-prefect/blob/main/internal/provider/resources/service_account.go#L331-L339

Reply:
That change results in:

```
$ terraform apply
...
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # prefect_block.secret must be replaced
-/+ resource "prefect_block" "secret" {
      ~ created   = "2024-06-07T16:32:08Z" -> (known after apply)
      ~ id        = "b015fc44-4a05-4f41-90b6-2d2523bb92bb" -> (known after apply)
        name      = "foo"
      ~ type_slug = "secret" -> "lambda" # forces replacement
      ~ updated   = "2024-06-07T16:32:08Z" -> (known after apply)
        # (1 unchanged attribute hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

What do you think of that flow @parkedwards? Or would you rather reject the change entirely?

Reply:
>  the high level is that you can just invoke req.Plan.Get() or req.State.Get()

🤦🏼 I tried that and my editor didn't want to autocomplete `.State` so I assumed it wasn't available. Must have had an error elsewhere in my file. Thanks!

The planmodifier approach _feels_ more "correct" here based on some googling but I'm also open to just grabbing `State` and `Plan` in the same place. Do you have a preference?

Reply:
I went with `planmodifier.RequiresReplace` in 5995859 for now as as it was the least amount of code, and I saw we use it in 2 other places in the codebase already. If we ever want to change this functionality to instead reject the change and throw an error, we know where to look.

Reply:
> should we allow users modifying the schema of the same type? im not sure we need to (eg. the UI does not expose this). also since we take the "[0]" schema from blockSchemaClient.List, we're effectively applying the latest schema on every update call

Fair question, probably not for now if the UI doesn't expose that. Although maybe it's worth a debug log of the block schema's metadata 🤔